asaplib/cluster/ml_cluster_fit.py

```diff
@@ -100,115 +100,267 @@ def pack(self):
+        -------
         """
-        plt.scatter(rho, delta)
-        plt.plot([min(rho), max(rho)], [self.deltamin, self.deltamin], c='red')
-        plt.plot([self.rhomin, self.rhomin], [min(delta), max(delta)], c='red')
-        plt.xlabel('rho')
-        plt.ylabel('delta')
-        plt.show()
-        ###
-
-        nclust = 0
-        cl = np.zeros(len(rho), dtype='int') - 1
-        for i in range(len(rho)):
-            # we select cluster centers {cl} with rho({c}) > rhomin && delta({c}) > delta_min
-            if rho[i] > self.rhomin and delta[i] > self.deltamin:
-                nclust += 1
-                cl[i] = nclust
-
-        ### Assignment
-        ordrho = np.argsort(rho)[::-1]
-        rho_ord = rho[ordrho]
-        for i in range(len(rho)):
-            if cl[ordrho[i]] == -1:
-                cl[ordrho[i]] = cl[nneigh[ordrho[i]]]
-        return cl
-
-    def estimate_delta(self, dist, rho):
-
-        ''' For each data point i, compute the distance (delta_i) between i and j,
-           j is the closet data point that has a density higher then i, i.e. rho(j) > rho(i).
-        '''
-
-        delta = (rho * 0.0).copy()
-        nneigh = np.ones(len(delta), dtype='int')
-        for i in range(len(rho)):
-            # for data i, find all points that have higher density
-            js = np.where(rho > rho[i])[0]
-            if len(js) == 0:
-                # if there's no j's that have higher density than i, we set delta_i to be a large distance
-                delta[i] = np.max(dist[i, :])
-                nneigh[i] = i
+        ordered = np.argsort(-self.dens)  # in descending order of density
+        self.cluster = np.zeros(data.shape[0], dtype='int')
+        tt = np.arange(data.shape[0])
+        center_label = np.zeros(data.shape[0], dtype='int')
+        ncluster = -1
+        for i in range(data.shape[0]):
+            j = ordered[i]
+            if (self.dens[j] > self.dens_cut) & (self.delta[j] > self.delta_cut):
+                ncluster = ncluster + 1
+                self.cluster[j] = ncluster
+                center_label[j] = ncluster
             else:
-                # find the nearest j that has higher density then i
-                delta[i] = np.min(dist[i, js])
-                nneigh[i] = js[np.argmin(dist[i, js])]
-        return delta, nneigh
+                self.cluster[j] = self.cluster[self.ref[j]]
+                center_label[j] = -1
+        self.centers = tt[(center_label != -1)]
+        bord = np.zeros(data.shape[0], dtype='int')
+        self.halo = np.copy(self.cluster)
+
+        for i in range(data.shape[0]):
+            for j in self.indices[i, :][(self.distances[i, :] <= self.dc)]:
+                if self.cluster[i] != self.cluster[j]:
+                    bord[i] = 1
+        halo_cutoff = np.zeros(ncluster + 1)
+        for i in range(ncluster + 1):
+            dd = self.dens[((bord == 1) & (self.cluster == i))]
+            halo_cutoff[i] = np.max(dd)
```

Comment:
This is the line that's currently failing with the Ni example. @BingqingCheng would be useful if you could reproduce the error.

`python clustering.py -kmat Ni-icosa-n6-l6-c4.0-g0.5.kmat --prefix Ni-icosa`

Reply:
This is now line 376. It is still failing with the Ni example. `dd` is an empty array. The class is very poorly documented so this needs to improve. @BingqingCheng where is the reference for this algorithm so that I can try to understand what the original author of this code intended?

Reply:
The halo computation logic is failing with the Ni example. `bord` is a binary array of length equal to the dataset size where a 1 indicates that there exist neighbouring data points (within the cutoff distance) belonging to a different cluster. The problem arises when considering the first cluster. In this cluster there are no border points and hence the expression `self.dens[((bord == 1) & (self.cluster == i))]` cannot be evaluated. This is a bug in the original code.

Reply:
@BingqingCheng I'll need to think further about how the halo should be defined for these clusters. Suggestions welcome

Reply:
I think I'm a bit lost. Can you be more specific about which algorithm are you referring to?

Just to be very clear, we want to be able to use the tools in the DPA external library.

Reply:
The method I'm referencing is the `get_assignation` method of the new `LAIO_DB` class which you asked me to replace your old class with. In order to do this, I want to ensure that the new class is compatible with your examples. The specific lines of code I'm referencing are next to this comment thread. The new class currently fails with the Ni example with the error described above relating to the halo computation. I believe (tentatively as it is challenging to understand the logic of the uncommented DPA code) that there is a bug in the DPA library algorithm when dealing with clusters which contain no halo points. One such cluster features in your Ni example. I will have to re-write the `get_assignation` method so that it doesn't malfunction for clusters containing no halo points. Your old class is included in the comments above beginning on line 105 for reference.

Zooming out for a moment, if instead of just replacing the `LAIO_DB` class, you'd like to support the DPA external library then I guess it's a question of whether you'd like the DPA library to work with your current examples of whether you'd like users to be able to write their own examples using the DPA library. In the latter case the halo computation bug is still worth raising with the maintainers.

 Is there a public repo for the DPA library? I tried Googling it but couldn't see anything obvious.

Reply:
Okay. Thanks for the explanations. Now I think I understand the issue. So basically a "feature" of the DPA implementation doesn't work for the Ni example because one cluster there has no halo points.

Do you think you can fix this?

I agree that we should write to the developers, either before or after you fixing this. You can write to Alex Rodriguez. I send you him email via slack.

If this is fixed, then I assume that your new LAIO_DB class works and if so we can just use it?